aws/rust-runtime/aws-http/src/recursion_detection.rs

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use crate::recursion_detection::env::TRACE_ID;
+use aws_smithy_http::middleware::MapRequest;
+use aws_smithy_http::operation::Request;
+use aws_types::os_shim_internal::Env;
+use http::HeaderValue;
+use percent_encoding::{percent_encode, CONTROLS};
+use std::borrow::Cow;
+
+/// Recursion Detection Middleware
+///
+/// This middleware inspects the value of the `AWS_LAMBDA_FUNCTION_NAME` and `_X_AMZ_TRACE_ID` environment
+/// variables to detect if the request is being invoked in a lambda function. If it is, the `X-Amzn-Trace-Id` header
+/// will be set. This enables downstream services to prevent accidentally infinitely recursive invocations spawned
+/// from lambda.
+#[non_exhaustive]
+#[derive(Default, Debug, Clone)]
+pub struct RecursionDetectionStage {
+    env: Env,
+}
+
+impl RecursionDetectionStage {
+    /// Creates a new `RecursionDetectionStage`
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl MapRequest for RecursionDetectionStage {
+    type Error = std::convert::Infallible;
+
+    fn apply(&self, request: Request) -> Result<Request, Self::Error> {
+        request.augment(|mut req, _conf| {
+            augument_request(&mut req, &self.env);
+            Ok(req)
+        })
+    }
+}
+
+const TRACE_ID_HEADER: &str = "x-amzn-trace-id";
+
+mod env {
+    pub(super) const LAMBDA_FUNCTION_NAME: &str = "AWS_LAMBDA_FUNCTION_NAME";
+    pub(super) const TRACE_ID: &str = "_X_AMZ_TRACE_ID";
+}
+
+/// Set the trace id header from the request
+fn augument_request<B>(req: &mut http::Request<B>, env: &Env) {
+    if req.headers().contains_key(TRACE_ID_HEADER) {
+        return;
+    }
+    if let (Ok(_function_name), Ok(trace_id)) =
+        (env.get(env::LAMBDA_FUNCTION_NAME), env.get(TRACE_ID))
```

Comment:
If you're just checking `LAMBDA_FUNCTION_NAME` to see if you're running in a Lambda, I believe checking `LAMBDA_TASK_ROOT` is preferred per this discussion https://github.com/awslabs/smithy-rs/pull/893#discussion_r755820444